src/modules/Playlists/XSPF.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+	QMPlay2 is a video and audio player.
+	Copyright (C) 2010-2016  Błażej Szczygieł
+
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Lesser General Public License as published
+	by the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Lesser General Public License for more details.
+
+	You should have received a copy of the GNU Lesser General Public License
+	along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <XSPF.hpp>
+
+#include <Functions.hpp>
+#include <Reader.hpp>
+#include <Writer.hpp>
+
+#include <QXmlStreamReader>
+#include <QXmlStreamWriter>
+#include <QFileInfo>
+
+/**/
+
+XSPF::XSPF(bool useRelative) :
+	m_useRelative(useRelative)
+{}
+XSPF::~XSPF()
+{}
+
+Playlist::Entries XSPF::read()
+{
+	Reader *reader = ioCtrl.rawPtr<Reader>();
+	const QString readFilePath = Functions::filePath(reader->getUrl().mid(7));
+	Entries list;
+
+	QXmlStreamReader xmlReader(reader->read(reader->size()));
+
+	bool inEntry = false;
+	Entry curr;
+	while (!xmlReader.atEnd() && !xmlReader.hasError())
+	{
+		QXmlStreamReader::TokenType token = xmlReader.readNext();
+		if (token == QXmlStreamReader::StartDocument)
+			continue;
+		if (token == QXmlStreamReader::StartElement)
+		{
+			const QStringRef name = xmlReader.name();
+			if (name == "track")
+			{
+				if (inEntry)
+				{
+					if (curr.name.isEmpty())
+						curr.name = Functions::fileName(curr.url, false);
+					list += curr;
+				}
+				curr = Entry();
+				inEntry = true;
+			}
+			else if (inEntry)
+			{
+				const QString value = xmlReader.readElementText();
+				if (name == "location")
+				{
+					if (value.startsWith("file://")) // local absolute path
+						curr.url = Functions::Url(value);
+					else if (Functions::splitPrefixAndUrlIfHasPluginPrefix(value, NULL, NULL, NULL)) // plugin URL
+						curr.url = value;
+					else if(QFileInfo::exists(readFilePath + value)) // relative path
+						curr.url = Functions::Url(value, readFilePath);
+				}
+				else if (name == "title")
+					curr.name = value;
+				else if (name == "duration")
+					curr.length = value.toInt() / 1000.0;
+				else if (name == "selected")
```

Comment:
"selected" is now obsolete. XSPF shouldn't have this key, because there is no need for compatibility :)